### PR TITLE
ref(ts) Convert GroupStore and StreamManager to typescript

### PR DESCRIPTION
**Pull request by markstory**
I had to do some additional checks as we were not handling all the undefined cases before.

### CODE REVIEW

**Comment by scttcper** [comment]
Oh nice GroupStore. I was just wishing that it was in typescript when working on something today.

**Review by scttcper — APPROVED** [review]

File: src/sentry/static/sentry/app/stores/groupStore.tsx [hunk 1]
```diff
@@ -2,58 +2,122 @@ import isArray from 'lodash/isArray';
...
 import {t} from 'app/locale';
 
 function showAlert(msg, type) {
-  IndicatorStore.add(msg, type, {
+  IndicatorStore.addMessage(msg, type, {
     duration: 4000,
   });
 }
 
+// TODO(ts) Type this any better.
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/stores/groupStore.tsx:16]: 😀

File: src/sentry/static/sentry/app/utils/streamManager.tsx [hunk 2]
```diff
@@ -6,10 +6,23 @@ const removeFromList = (item, list) => {
   }
 };
 
+type Options = {
+  limit?: number;
+};
+
+type IdThing = {
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/utils/streamManager.tsx:13]: are you sure about this name?

File: src/sentry/static/sentry/app/stores/groupStore.tsx [hunk 3]
```diff
@@ -2,58 +2,122 @@ import isArray from 'lodash/isArray';
...
-  remove(changeId, itemId) {
+  remove(changeId: string, itemId?: string) {
     this.changes = this.changes.filter(
       change => change[0] !== changeId || change[1] !== itemId
     );
   }
 
-  forEach() {
-    this.changes.forEach.apply(this.changes, arguments);
+  forEach(...args) {
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/stores/groupStore.tsx:36]:
```suggestion
  forEach(...args: any[]) {
```

File: src/sentry/static/sentry/app/utils/streamManager.tsx (hunk 2 above)
Comment (markstory) [reply src/sentry/static/sentry/app/utils/streamManager.tsx:13]: Would `IdShape` be better?

File: src/sentry/static/sentry/app/stores/groupStore.tsx [hunk 4]
```diff
@@ -2,58 +2,114 @@ import isArray from 'lodash/isArray';
...
 
 function showAlert(msg, type) {
-  IndicatorStore.add(msg, type, {
+  IndicatorStore.addMessage(msg, type, {
     duration: 4000,
   });
 }
 
+// TODO(ts) Type this any better.
+type Change = [string, string, any];
```
Comment (k-fish) [inline src/sentry/static/sentry/app/stores/groupStore.tsx:17]: As of typescript 4 you can do named tuples 😄, not like it helps a ton but it's nice to know which of these values is supposed to be which on hover. The nice thing about this is you can replace rest params with a named tuple and the params will still be properly labeled.
```suggestion
type Change = [changeId: string, itemId:string, data:any];
```

File: src/sentry/static/sentry/app/stores/groupStore.tsx [hunk 5]
```diff
@@ -2,58 +2,114 @@ import isArray from 'lodash/isArray';
...
 
-  getForItem(itemId) {
-    return this.changes.filter(change => change[1] === itemId);
+  getForItem(itemId: string) {
+    return this.changes.filter((change: Change) => change[1] === itemId);
   }
 
-  push(changeId, itemId, data) {
+  push(changeId: string, itemId: string, data: any) {
     this.changes.push([changeId, itemId, data]);
```
Comment (k-fish) [inline src/sentry/static/sentry/app/stores/groupStore.tsx:27]:
```suggestion
  push(...change: Change) {
    this.changes.push(change);
```

File: src/sentry/static/sentry/app/stores/groupStore.tsx [hunk 6]
```diff
@@ -2,58 +2,114 @@ import isArray from 'lodash/isArray';
...
+  loadInitialData: (items: Group[]) => void;
+  add: (items: Group[]) => void;
+  remove: (itemId: string) => void;
+  addStatus: (id: string, status: string) => void;
+  clearStatus: (id: string, status: string) => void;
+  hasStatus: (id: string, status: string) => boolean;
+  get: (id: string) => Group | undefined;
+  getAllItemIds: () => string[];
+  getAllItems: () => Group[];
+  onAssignTo: (changeId: string, itemId: string, data: any) => void;
```
Comment (k-fish) [inline src/sentry/static/sentry/app/stores/groupStore.tsx:64]: Following along from above

```suggestion
  onAssignTo: (...change: Change) => void;
```

**Review by k-fish — APPROVED** [review]
lgtm, named tuples is just a suggestion.

File: src/sentry/static/sentry/app/utils/streamManager.tsx (hunk 2 above)
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/utils/streamManager.tsx]: yes :)

**Review by priscilawebdev — APPROVED** [review]

**Pull request merged**